pkg/client/introduce/client.go

```diff
@@ -9,220 +7,131 @@ SPDX-License-Identifier: Apache-2.0
-// It can be executed after receiving a Request action message (the client does not have a public Invitation)
-func (c *Client) HandleRequest(msg service.DIDCommMsg, to *introduce.To, recipient *introduce.Recipient) error {
-	thID, err := msg.ThreadID()
-	if err != nil {
-		return fmt.Errorf("handle request threadID: %w", err)
-	}
-
-	return c.saveInvitationEnvelope(thID, InvitationEnvelope{
-		Recps: []*introduce.Recipient{{To: to}, recipient},
-	})
-}
-
-// HandleRequestWithInvitation is a helper function to prepare the right protocol dependency interface.
-// It can be executed after receiving a Request action message (the client has a public Invitation)
-func (c *Client) HandleRequestWithInvitation(msg service.DIDCommMsg,
-	inv *didexchange.Invitation, to *introduce.To) error {
-	thID, err := msg.ThreadID()
-	if err != nil {
-		return fmt.Errorf("handle request with invitation threadID: %w", err)
-	}
+	}), myDID, theirDID)
 
-	return c.saveInvitationEnvelope(thID, InvitationEnvelope{
-		Inv:   inv,
-		Recps: []*introduce.Recipient{{To: to}},
-	})
+	return err
 }
 
-func (c *Client) handleOutbound(msg interface{}, o InvitationEnvelope) error {
-	payload, err := json.Marshal(msg)
-	if err != nil {
-		return fmt.Errorf("marshal outbound msg: %w", err)
-	}
-
-	didMsg, err := service.ParseDIDCommMsgMap(payload)
-	if err != nil {
-		return fmt.Errorf("new outbound DIDCommMsg msg: %w", err)
-	}
-
-	thID, err := didMsg.ThreadID()
-	if err != nil {
-		return fmt.Errorf("outbound threadID: %w", err)
-	}
-
-	if _, ok := msg.(*introduce.Request); !ok {
-		if err := c.saveInvitationEnvelope(thID, o); err != nil {
-			return err
-		}
-	}
-
-	return c.service.HandleOutbound(didMsg, o.Recps[0].MyDID, o.Recps[0].TheirDID)
-}
-
-// InvitationEnvelope keeps the information needed for sending a proposal
-type InvitationEnvelope struct {
-	// Invitation must be set when we have a public Invitation
-	Inv *didexchange.Invitation `json:"invitation,omitempty"`
-	// Recipients contain one or two elements
-	// one element - for the public Invitation, otherwise two elements
-	Recps []*introduce.Recipient `json:"recipients,omitempty"`
+// ContinueWithInvitation is used when introducee wants to provide invitation.
+// NOTE: For async usage. Introducee can provide invitation only after receiving ProposalMsgType
+func (c *Client) ContinueWithInvitation(piID string, inv *didexchange.Invitation) error {
+	return c.service.Continue(piID, WithInvitation(inv))
 }
 
-// Invitation returns an Invitation needed for the service (state machine depends on it)
-func (o *InvitationEnvelope) Invitation() *didexchange.Invitation {
-	return o.Inv
+// ContinueWithPublicInvitation is used when introducer wants to provide public invitation.
+// NOTE: For async usage. Introducer can provide invitation only after receiving RequestMsgType
+func (c *Client) ContinueWithPublicInvitation(piID string, inv *didexchange.Invitation, to *introduce.To) error {
```

Comment:
I'm a bit confused. Why are we exposing both `ContinueWithPublicInvitation` and `WithPublicInvitation` from this package? And we haven't exposed a generic `Continue` for `WithPublicInvitation`?

Reply:
ContinueWithPublicInvitation for async. WithPublicInvitation is an option for the Continue function (when we received an event message)

Reply:
This is too confusing.

Reply:
When you received an action event you can call a.Continue(WithPublicInvitation) . It is much cheaper because we avoiding the call to DB. 

When you call API  ContinueWithPublicInvitation(PIID,..) It will get data from the DB (by PIID)

Reply:
I also don't see this happening in DID Exchange client: https://godoc.org/github.com/hyperledger/aries-framework-go/pkg/client/didexchange

Reply:
I think we didn't wrap the event continuation functions - as these are considered advanced usage... instead, we simply created client methods for the more typical async case. 

@rolsonquadras @llorllale @soluchok

Reply:
In DID Exchange it is `AcceptExchangeRequest` function

Reply:
I think to reduce confusion, I see two possibilities:
- Supply individual client methods for the async case but no wrappers for the event continuation functions.
- Supply wrappers for continuation functions and only have a single Continue method in the client.

I think DID Exchange client has the first mechanism.

Reply:
@troyronda agree. The DIDExchange client/service has individual APIs and handle() function is called from these APIs inside service pkg (e.Continue calls handle() as well). We can wrap svc.Continue() function in client package as well. The idea in DIDExchange was to give a clear APIs to the consumers, so that they would know what exact action they are performing.

Reply:
@rolsonquadras ok. so how should this PR be changed?

Reply:
@troyronda I would suggest associating the function names with action/state(stateMachine) rather than input.

- introduce.AcceptProposal(piID string, inv *didexchange.Invitation)
- introduce.AcceptRequestWithPublicInvitation(piID string, inv *didexchange.Invitation, to *To)
- introduce.AcceptRequestWithRecipients(piID string, to *To, recipient *Recipient)
- introduce.Actions()

Reply:
> When you received an action event you can call a.Continue(WithPublicInvitation) . It is much cheaper because we avoiding the call to DB.
> 
> When you call API ContinueWithPublicInvitation(PIID,..) It will get data from the DB (by PIID)

`Continue()` is hitting the store always.

Reply:
@troyronda 

> IMO - this is confusing. Especially since ContinueWithPublicInvitation is basically just a wrapper for Continue + WithPublicInvitation.

Agreed

> I think we didn't wrap the event continuation functions in DID Exchange - as these are considered advanced usage... instead, we simply created client methods for the more typical async case.

I consider the "continue with this invitation" capability required for basic use of the introduce protocol.

> * Supply individual client methods for the async case but no wrappers for the event continuation functions.
> * Supply wrappers for continuation functions and only have a single Continue method in the client.

I think the second option would look more elegant.

Reply:
> I consider the "continue with this invitation" capability required for basic use of the introduce protocol 

It's not basic usage to do so from an event object.